src/lib/shady-render.ts

```diff
@@ -272,12 +273,11 @@ export const render =
       if (firstScopeRender) {
         const part = parts.get(renderContainer)!;
         parts.delete(renderContainer);
-        if (part.value instanceof TemplateInstance) {
-          prepareTemplateStyles(
-              renderContainer as DocumentFragment,
-              part.value.template,
-              scopeName);
-        }
+        const template = part.value instanceof TemplateInstance ?
+            part.value.template :
+            new Template(html``, document.createElement('template'));
```

Comment:
What's this part for? Add a comment.

Reply:
Added, let me know what you think.

Reply:
Is this really going to work? `prepareTemplateStyles` might insert styles into this template, but for what purpose if the template is never rendered anywhere?

I get the feeling that this isn't going to work with the original use-case of rendering `undefined`, then rendering a `TemplateResult`. Can we add a test specifically for that, using shady-render?

Reply:
I think the gist is that the styling we are about in this case was sent to ShadyCSS via `prepareAdoptedCssText`. Styling in the non-template result would just not be supported.

So it seems like all we need to do in this case is ensure that (1) styles are removed from any lit template for the scope (would there be any if no TemplateResult is being rendered?, (2) ensure that `prepareTemplate` is called (because this is necessary to apply any styles defined with `prepareAdoptedCssText`).  Why not change `prepareTemplateStyles` to optionally take a Template rather than creating a dummy one here?

Reply:
> Is this really going to work? `prepareTemplateStyles` might insert styles into this template, but for what purpose if the template is never rendered anywhere?
>
> I get the feeling that this isn't going to work with the original use-case of rendering `undefined`, then rendering a `TemplateResult`. Can we add a test specifically for that, using shady-render?

Ok, I added another test specifically for this case.

> I think the gist is that the styling we are about in this case was sent to ShadyCSS via `prepareAdoptedCssText`. Styling in the non-template result would just not be supported.

AFAICT, the current setup seems to be that [all styles in the first render are scoped](https://github.com/Polymer/lit-html/blob/f37bf689fb836a009f1a6a413bc9e7d7262964ae/src/lib/shady-render.ts#L131) and [any styles that appear later aren't](https://github.com/Polymer/lit-html/blob/f37bf689fb836a009f1a6a413bc9e7d7262964ae/src/lib/shady-render.ts#L121-L125). So I think that even styles inserted into NodeParts of a top-level TemplateInstance were collected / scoped / removed from their rendered position on the first render.

Am I correct in thinking that (local) `prepareTemplateStyles`' `renderedDOM` contains the complete rendered output of the first render by the time it runs?

> So it seems like all we need to do in this case is ensure that (1) styles are removed from any lit template for the scope (would there be any if no TemplateResult is being rendered?, (2) ensure that `prepareTemplate` is called (because this is necessary to apply any styles defined with `prepareAdoptedCssText`). Why not change `prepareTemplateStyles` to optionally take a Template rather than creating a dummy one here?

I was avoiding this because it seemed difficult to pry apart the Template editing parts of (local) `prepareTemplateStyles` from the parts that were dealing with changing the output of the first render but it seems necessary at this point to match the earlier behavior - I'll update.

Reply:
I made `prepareTemplateStyles`' `template` parameter optional and updated the logic in there to work with a new, empty template if it isn't passed. Also, I added a couple extra tests that ShadyCSS mixins and scoping are applied to styles that are part of an initial render for a scope of a non-TemplateInstance value.